File: sdk-swift/Sources/PlaidifyLinkKit/PlaidifyLinkConnectFlow.swift
import Foundation

/// Step a hosted-link flow is currently on. Mirrors the React app's
/// reducer so embedders see a consistent state machine.
public enum PlaidifyLinkStep: String, Equatable {
    case consent
    case picker
    case credentials
    case connecting
    case mfa
    case success
    case error
}

/// A single observed callback emitted by the flow. Callers translate
/// these into native UI updates and/or webview bridge events.
public enum PlaidifyLinkFlowEvent: Equatable {
    case stepChanged(PlaidifyLinkStep)
    case institutionSelected(PlaidifyOrganization)
    case mfaRequired(type: String, sessionID: String?)
    case connected(publicToken: String?, jobID: String?, site: String?)
    case errored(code: String?, message: String)
    case fallbackToWebView(PlaidifyOrganization, reason: String)
}

/// Pure state for the connect flow. Holds no UI dependencies so it
/// can be unit-tested under `swift test` without simulator/device.
public struct PlaidifyLinkFlowState: Equatable {
    public var step: PlaidifyLinkStep
    public var organization: PlaidifyOrganization?
    public var sessionID: String?
    public var mfaType: String?
    public var lastErrorCode: String?
    public var lastErrorMessage: String?
    public var publicToken: String?
    public var jobID: String?

    public init(
        step: PlaidifyLinkStep = .picker,
        organization: PlaidifyOrganization? = nil,
        sessionID: String? = nil,
        mfaType: String? = nil,
        lastErrorCode: String? = nil,
        lastErrorMessage: String? = nil,
        publicToken: String? = nil,
        jobID: String? = nil
    ) {
        self.step = step
        self.organization = organization
        self.sessionID = sessionID
        self.mfaType = mfaType
        self.lastErrorCode = lastErrorCode
        self.lastErrorMessage = lastErrorMessage
        self.publicToken = publicToken
        self.jobID = jobID
    }
}

/// Coordinates the picker → credentials → MFA → success transitions.
///
/// The flow is driven by `apply` calls so a UI layer (SwiftUI/UIKit)
/// can dispatch user actions and observe the resulting events without
/// owning any business logic.
public final class PlaidifyLinkConnectFlow {
    public private(set) var state: PlaidifyLinkFlowState
    public let registry: PlaidifyLinkInstitutionRegistry
    public var onEvent: (PlaidifyLinkFlowEvent) -> Void

    public init(
        registry: PlaidifyLinkInstitutionRegistry = PlaidifyLinkInstitutionRegistry(),
        initialState: PlaidifyLinkFlowState = PlaidifyLinkFlowState(),
        onEvent: @escaping (PlaidifyLinkFlowEvent) -> Void = { _ in }
    ) {
        self.registry = registry
        self.state = initialState
        self.onEvent = onEvent
    }

    public enum Action: Equatable {
        case selectInstitution(PlaidifyOrganization)
        case credentialsSubmitted
        case connectResponded(PlaidifyConnectResponse)
        case mfaSubmitted
        case mfaResponded(PlaidifyConnectResponse)
        case failed(code: String?, message: String)
        case reset
    }

    @discardableResult
    public func apply(_ action: Action) -> PlaidifyLinkFlowState {
        switch action {
        case .selectInstitution(let organization):
            switch registry.strategy(for: organization) {
            case .webViewFallback(let reason):
                onEvent(.fallbackToWebView(organization, reason: reason))
                state.organization = organization
            case .native:
                state.organization = organization
                state.step = .credentials
                onEvent(.institutionSelected(organization))
                onEvent(.stepChanged(.credentials))
            }

        case .credentialsSubmitted:
            state.step = .connecting
            onEvent(.stepChanged(.connecting))

        case .connectResponded(let response):
            handleConnectResponse(response)

        case .mfaSubmitted:
            state.step = .connecting
            onEvent(.stepChanged(.connecting))

        case .mfaResponded(let response):
            handleConnectResponse(response)

        case .failed(let code, let message):
            state.step = .error
            state.lastErrorCode = code
            state.lastErrorMessage = message
            onEvent(.errored(code: code, message: message))
            onEvent(.stepChanged(.error))

        case .reset:
            state = PlaidifyLinkFlowState()
            onEvent(.stepChanged(.picker))
        }
        return state
    }

    private func handleConnectResponse(_ response: PlaidifyConnectResponse) {
        switch response.status {
        case "completed":
            state.step = .success
            state.publicToken = response.publicToken
            state.jobID = response.jobID
            onEvent(.connected(
                publicToken: response.publicToken,
                jobID: response.jobID,
                site: state.organization?.site
            ))
            onEvent(.stepChanged(.success))

        case "mfa_required":
            state.step = .mfa
            state.sessionID = response.sessionID
            state.mfaType = response.mfaType
            onEvent(.mfaRequired(type: response.mfaType ?? "otp", sessionID: response.sessionID))
            onEvent(.stepChanged(.mfa))

        case "error":
            apply(.failed(code: nil, message: response.errorMessage ?? response.message ?? "Connection failed."))

        default:
            // pending / unknown: stay on connecting
            break
        }
    }
}

File: sdk-android/core/src/main/kotlin/com/plaidify/link/PlaidifyLinkConnectFlow.kt
package com.plaidify.link

/** Step the hosted-link flow is currently on. */
public enum class PlaidifyLinkStep {
    Consent, Picker, Credentials, Connecting, Mfa, Success, Error
}

/** Single observable callback emitted by [PlaidifyLinkConnectFlow]. */
public sealed class PlaidifyLinkFlowEvent {
    public data class StepChanged(val step: PlaidifyLinkStep) : PlaidifyLinkFlowEvent()
    public data class InstitutionSelected(val organization: PlaidifyOrganization) : PlaidifyLinkFlowEvent()
    public data class MfaRequired(val type: String, val sessionId: String?) : PlaidifyLinkFlowEvent()
    public data class Connected(
        val publicToken: String?,
        val jobId: String?,
        val site: String?,
    ) : PlaidifyLinkFlowEvent()

    public data class Errored(val code: String?, val message: String) : PlaidifyLinkFlowEvent()
    public data class FallbackToWebView(
        val organization: PlaidifyOrganization,
        val reason: String,
    ) : PlaidifyLinkFlowEvent()
}

/** Pure state for the connect flow. */
public data class PlaidifyLinkFlowState(
    val step: PlaidifyLinkStep = PlaidifyLinkStep.Picker,
    val organization: PlaidifyOrganization? = null,
    val sessionId: String? = null,
    val mfaType: String? = null,
    val lastErrorCode: String? = null,
    val lastErrorMessage: String? = null,
    val publicToken: String? = null,
    val jobId: String? = null,
)

/**
 * Pure state machine for the picker → credentials → connecting →
 * mfa → success/error transitions. The class is intentionally
 * UI-framework agnostic so it is fully unit-testable on the JVM.
 */
public class PlaidifyLinkConnectFlow(
    public val registry: PlaidifyLinkInstitutionRegistry = PlaidifyLinkInstitutionRegistry(),
    initialState: PlaidifyLinkFlowState = PlaidifyLinkFlowState(),
    public var onEvent: (PlaidifyLinkFlowEvent) -> Unit = {},
) {
    public var state: PlaidifyLinkFlowState = initialState
        private set

    public sealed class Action {
        public data class SelectInstitution(val organization: PlaidifyOrganization) : Action()
        public object CredentialsSubmitted : Action()
        public data class ConnectResponded(val response: PlaidifyConnectResponse) : Action()
        public object MfaSubmitted : Action()
        public data class MfaResponded(val response: PlaidifyConnectResponse) : Action()
        public data class Failed(val code: String?, val message: String) : Action()
        public object Reset : Action()
    }

    public fun apply(action: Action): PlaidifyLinkFlowState {
        when (action) {
            is Action.SelectInstitution -> {
                when (val strategy = registry.strategy(action.organization)) {
                    is PlaidifyLinkInstitutionStrategy.WebViewFallback -> {
                        onEvent(PlaidifyLinkFlowEvent.FallbackToWebView(action.organization, strategy.reason))
                        state = state.copy(organization = action.organization)
                    }
                    PlaidifyLinkInstitutionStrategy.Native -> {
                        state = state.copy(
                            organization = action.organization,
                            step = PlaidifyLinkStep.Credentials,
                        )
                        onEvent(PlaidifyLinkFlowEvent.InstitutionSelected(action.organization))
                        onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Credentials))
                    }
                }
            }

            Action.CredentialsSubmitted -> {
                state = state.copy(step = PlaidifyLinkStep.Connecting)
                onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Connecting))
            }

            is Action.ConnectResponded -> handleConnectResponse(action.response)

            Action.MfaSubmitted -> {
                state = state.copy(step = PlaidifyLinkStep.Connecting)
                onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Connecting))
            }

            is Action.MfaResponded -> handleConnectResponse(action.response)

            is Action.Failed -> {
                state = state.copy(
                    step = PlaidifyLinkStep.Error,
                    lastErrorCode = action.code,
                    lastErrorMessage = action.message,
                )
                onEvent(PlaidifyLinkFlowEvent.Errored(action.code, action.message))
                onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Error))
            }

            Action.Reset -> {
                state = PlaidifyLinkFlowState()
                onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Picker))
            }
        }
        return state
    }

    private fun handleConnectResponse(response: PlaidifyConnectResponse) {
        when (response.status) {
            "completed" -> {
                state = state.copy(
                    step = PlaidifyLinkStep.Success,
                    publicToken = response.publicToken,
                    jobId = response.jobId,
                )
                onEvent(
                    PlaidifyLinkFlowEvent.Connected(
                        publicToken = response.publicToken,
                        jobId = response.jobId,
                        site = state.organization?.site,
                    )
                )
                onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Success))
            }

            "mfa_required" -> {
                state = state.copy(
                    step = PlaidifyLinkStep.Mfa,
                    sessionId = response.sessionId,
                    mfaType = response.mfaType,
                )
                onEvent(
                    PlaidifyLinkFlowEvent.MfaRequired(
                        type = response.mfaType ?: "otp",
                        sessionId = response.sessionId,
                    )
                )
                onEvent(PlaidifyLinkFlowEvent.StepChanged(PlaidifyLinkStep.Mfa))
            }

            "error" -> {
                apply(
                    Action.Failed(
                        code = null,
                        message = response.errorMessage ?: response.message ?: "Connection failed.",
                    )
                )
            }

            else -> {
                // pending / unknown: stay on connecting.
            }
        }
    }
}
